### PR TITLE
fix(repository): support embedded models for nested properties 

### LINKED ISSUE #2130: Support embedded models for nested properties
## Description / Steps to reproduce / Feature proposal
This issue was discussed at [stackoverflow](https://stackoverflow.com/questions/53554400/exception-in-deserialization-of-nested-objects-for-datamodels-in-loopback-4/). 

An error occurs on create if nested entities are defined. 

## Current Behavior

on create of an entity with a nested child (or more), an error occurs: 

> Unhandled error in POST /myMainEntity: 500 TypeError: Class constructor MySubEntity cannot be invoked without 'new'

if the nested entity was included like this:
```
@property({
    type: MySubEntity, // maybe in aphostrope
  })
  subEntity?: MySubEntity;
```

## Expected Behavior
The objective of this was to get nested entities, which
- are type safe (checks on the properties of the sub object; autocomplete on the sub-object; swagger shows the correct sub-object; ...)
- are invoked correctly with their containing datatypes (e.g. transformation from String to Date if the property is defined as a date)

_Even more helpful would be to use jsonSchema for the datamodel and get the described features like autocomplete, typesafety if this is possible anyhow._

### PR DESCRIPTION
**Pull request by IsCaster**
Fixes #2130

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

File: packages/repository/src/__tests__/unit/repositories/legacy-juggler-bridge.unit.ts [hunk 1]
```diff
@@ -12,6 +12,8 @@ import {
   juggler,
   ModelDefinition,
 } from '../../..';
+import {Model} from '../../../model';
+import {property, model} from '../../../decorators';
```
Comment (bajtos) [inline packages/repository/src/__tests__/unit/repositories/legacy-juggler-bridge.unit.ts]: Let's keep the imports simple and import everything from `../../..`.

File: packages/repository/src/repositories/legacy-juggler-bridge.ts [hunk 2]
```diff
@@ -93,26 +97,53 @@ export class DefaultCrudRepository<T extends Entity, ID>
   ) {
     const definition = entityClass.definition;
     assert(
-      !!definition,
-      `Entity ${entityClass.name} must have valid model definition.`,
```
Comment (bajtos) [inline packages/repository/src/repositories/legacy-juggler-bridge.ts]: Please preserve this assert and add your new assertion as a new `assert` call.

When `definition` is undefined, your check `definition.idProperties().length > 0` will crash on trying to access `idProperties` on an undefined object.

File: packages/repository/src/repositories/legacy-juggler-bridge.ts [hunk 3]
```diff
@@ -93,26 +97,53 @@ export class DefaultCrudRepository<T extends Entity, ID>
...
+      definition.idProperties().length > 0,
+      `Entity ${entityClass.name} must have at least one id/pk property.`,
     );
 
+    this.setupModel(entityClass);
+    this.setModelClass();
+  }
+  private setupModel(entityClass: typeof Model & {prototype: T}) {
```
Comment (bajtos) [inline packages/repository/src/repositories/legacy-juggler-bridge.ts]: We are adding an empty line between methods to visually delimit them.

```ts
}

private setupModel(entityClass: typeof Model & {prototype: T}) {
```

File: packages/repository/src/repositories/legacy-juggler-bridge.ts [hunk 4]
```diff
@@ -93,26 +97,53 @@ export class DefaultCrudRepository<T extends Entity, ID>
...
+    const definition = entityClass.definition;
+
     assert(
-      definition.idProperties().length > 0,
-      `Entity ${entityClass.name} must have at least one id/pk property.`,
+      !!definition,
```
Comment (bajtos) [inline packages/repository/src/repositories/legacy-juggler-bridge.ts]: As I wrote above, this check should be run as the first one.

File: packages/repository/src/repositories/legacy-juggler-bridge.ts [hunk 5]
```diff
@@ -93,26 +97,53 @@ export class DefaultCrudRepository<T extends Entity, ID>
...
+      }
+    }
+
+    this.attachModelToDataSource(definition);
+  }
+
+  private setModelClass() {
+    const model = this.dataSource.getModel(this.entityClass.definition.name);
+    if (model) {
+      // The backing persisted model has been already defined.
```
Comment (bajtos) [inline packages/repository/src/repositories/legacy-juggler-bridge.ts]: What if it has not been registered? 

I think the following code may work well:

```ts
const model = this.dataSource.getModel(this.entityClass.definition.name);
assert(!!model, 'The backing persisted model should have been defined by setupModel()');
this.modelClass = model as typeof juggler.PersistedModel;
```

Even better, can we modify `setupModel` to return the PersistedModel class created? The we could change the constructor as follows:

```ts
this.modelClass = this.setupModel(entityClass);
```

I would also rename `setupModel` to something like `definePersistedModel`.

File: packages/repository/src/repositories/legacy-juggler-bridge.ts [hunk 6]
```diff
@@ -93,26 +97,53 @@ export class DefaultCrudRepository<T extends Entity, ID>
...
   }
 
   // Create an internal legacy Model attached to the datasource
-  private setupPersistedModel(definition: ModelDefinition) {
+  private attachModelToDataSource(definition: ModelDefinition) {
     const dataSource = this.dataSource;
 
     const model = dataSource.getModel(definition.name);
     if (model) {
-      // The backing persisted model has been already defined.
```
Comment (bajtos) [inline packages/repository/src/repositories/legacy-juggler-bridge.ts]: Please preserve this comment.

File: packages/repository/src/repositories/legacy-juggler-bridge.ts (hunk 4 above)
Comment (IsCaster) [reply packages/repository/src/repositories/legacy-juggler-bridge.ts]: This check also runs when it's a embedded model.

**Comment by IsCaster** [comment]
@bajtos @raymondfeng Hi, I had already changed the code. Could you merge it, please?

**Comment by bajtos** [comment]
@IsCaster sorry for the delays. I'll try to take a look at the proposed changes again by the end of this week.

**Comment by bajtos** [comment]
As I was reviewing the changes in more details, I realize we need to improve the implementation and decided it will be faster for me to make the changes rather than try to explain them. See 45e88ce.

I have also rebased the pull request on top of the latest master.

@IsCaster please take a look, are you ok with the changes I made?

File: packages/repository/src/repositories/legacy-juggler-bridge.ts [hunk 7]
```diff
@@ -124,14 +141,15 @@ export class DefaultCrudRepository<T extends Entity, ID>
     // the juggler understands
     Object.entries(definition.properties).forEach(([key, value]) => {
       if (value.type === 'array' || value.type === Array) {
-        value = Object.assign({}, value, {type: [resolveType(value.itemType)]});
+        value = Object.assign({}, value, {
+          type: [value.itemType && this.resolvePropertyType(value.itemType)],
```
Comment (bajtos) [inline packages/repository/src/repositories/legacy-juggler-bridge.ts]: We want juggler to see the new persisted model in the property definition, not a reference to LB4 model.

File: packages/repository/src/repositories/legacy-juggler-bridge.ts [hunk 8]
```diff
@@ -124,14 +141,15 @@ export class DefaultCrudRepository<T extends Entity, ID>
...
+        });
         delete value.itemType;
       }
-      value.type = resolveType(value.type);
+      value.type = this.resolvePropertyType(value.type);
```
Comment (bajtos) [inline packages/repository/src/repositories/legacy-juggler-bridge.ts]: Similarly here: We want juggler to see the new persisted model in the property definition, not a reference to LB4 model.

**Comment by IsCaster** [comment]
> As I was reviewing the changes in more details, I realize we need to improve the implementation and decided it will be faster for me to make the changes rather than try to explain them. See [45e88ce](https://github.com/strongloop/loopback-next/commit/45e88ce994ed734988ebecc34fc9ff72fd9f440c).
> 
> I have also rebased the pull request on top of the latest master.
> 
> @IsCaster please take a look, are you ok with the changes I made?

Ok with me

**Comment by bajtos** [comment]
Landed, thank you @IsCaster for the contribution ❤️

**Pull request merged**